src/service/rbac-service.ts

```diff
@@ -17,42 +17,246 @@
+
+          entDef[permission2.action] = entityPermissions
+            .filter((p) => p.action === permission2.action)
+            .reduce((actDef: ActionDefinition, permission3) => {
+              if (actDef[permission3.relation]) return;
+
+              actDef[permission3.relation] = new Set(permission3.attributes);
+              return actDef;
+
+            }, {});
+          return entDef;
+
+        }, {});
+      return permDef;
+    }, {});
+  }
+
+  /**
+   * Get a tuple with a list of all roles and the total number of
+   * roles matching the parameters
+   * @param params
+   * @param take
+   * @param skip
+   */
+  public static async getRoles(params: RoleFilterParameters = {}, { take, skip }: PaginationParameters = {}): Promise<[Role[], number]> {
+    const options = this.getOptions(params);
+    return Role.findAndCount({ ...options, take, skip });
+  }
+
+  /**
+   * Create an new role with the given parameters
+   * @param params
+   */
+  public static async createRole(params: UpdateRoleParams) {
+    return Role.save({ ...params, permissions: [] });
+  }
+
+  /**
+   * Update an existing role with the given parameters
+   * @param roleId
+   * @param params
+   */
+  public static async updateRole(roleId: number, params: UpdateRoleParams): Promise<Role> {
+    const role = await Role.findOne({ where: { id: roleId } });
+    if (role == null) throw new Error('Role not found.');
+    if (role.systemDefault) throw new Error('Cannot update system default role.');
+
+    role.name = params.name;
+    await role.save();
+
+    const [[r]] = await this.getRoles({ roleId, returnPermissions: true });
+    return r;
+  }
+
+  /**
+   * Remove an existing role. Cannot delete system default roles
+   * @param roleId
+   */
+  public static async removeRole(roleId: number) {
+    const [[role]] = await this.getRoles({ roleId });
+    if (!role) throw new Error('Role not found.');
+    if (role.systemDefault) throw new Error('Cannot delete system default role.');
+
+    await Role.remove(role);
+  }
+
+  /**
+   * Add zero or more new permissions
+   * @param roleId
+   * @param permissions
+   */
+  public static async addPermissions(roleId: number, permissions: PermissionRule[]) {
+    const [[role]] = await this.getRoles({ roleId, returnPermissions: true });
+    if (!role) throw new Error('Role not found.');
+
+    return Permission.save(permissions.map((p): DeepPartial<Permission> => ({
+      ...p,
+      role,
+      roleId: role.id,
+    })));
+  }
+
+  /**
+   * Remove an existing permission
+   * @param roleId
+   * @param permissionRule
+   */
+  public static async removePermission(roleId: number, permissionRule: Omit<PermissionRule, 'attributes'>) {
+    const matches = (await Permission.find({ where: {
+      roleId,
+      entity: permissionRule.entity,
+      action: permissionRule.action,
+      relation: permissionRule.relation,
+    }, relations: { role: true } }));
+    if (matches.length === 0) {
```

Comment:
maybe then assign it to a const called `match`, I find array indexing, even if its with 0, to be ugly and prefer to swap it out as early as possible. I.e prefer a const instead of `matches[0]`

Reply:
Changed it!